include/lsst/afw/detection/Footprint.h

```diff
@@ -0,0 +1,383 @@
+     *
+     * @param key - A key corresponding to the field in the Scheam the PeakCatalog is to be
+     *              sorted by.
+     */
+    void sortPeaks(afw::table::Key<float> const & key=afw::table::Key<float>());
+
+    /** @brief Set the Schema used by the PeakCatalog (will throw if PeakCatalog is not empty).
+     *
+     * This function will throw an error if the PeakCatalog is not empty
+     *
+     * @param peakSchema - The schema to use in the PeakCatalog
+     */
+    void setPeakSchema(afw::table::Schema const & peakSchema) {
+        if (!getPeaks().empty()) {
+            throw LSST_EXCEPT(
+                pex::exceptions::LogicError,
+                "Cannot change the PeakCatalog schema unless it is empty"
+            );
+        }
+        // this syntax doesn't work in Python, which is why this method has to exist
+        getPeaks() = PeakCatalog(peakSchema);
+    }
+
+    /** @brief Set the peakCatalog to a copy of the supplied catalog
+     *
+     * PeakCatalog will be copied into the Footprint, but a PeakCatalog is a shallow
+     * copy, so records will not be duplicated. This function will throw an error if
+     * the PeakCatalog of *this is not empty.
+     *
+     * @param otherPeaks The PeakCatalog to copy
+     */
+    void setPeakCatalog(PeakCatalog const & otherPeaks) {
+        if (!getPeaks().empty()) {
+            throw LSST_EXCEPT(
+                pex::exceptions::LogicError,
+                "Cannot change the PeakCatalog unless it is empty"
+            );
+        }
+        getPeaks() = otherPeaks;
+    }
+
+    /**
+     * @brief Return the number of pixels in this Footprint
+     *
+     * This function returns the real number of pixels, not the area of the bbox.
+     */
+    size_t getArea() const { return _spans->getArea(); }
+
+    /**
+     * @brief Return the Footprint's centroid
+     *
+     * The centroid is calculated as the mean of the pixel centers
+     */
+    geom::Point2D getCentroid() const { return _spans->computeCentroid(); }
+
+    /**
+     * @brief Return the Footprint's shape (interpreted as an ellipse)
+     *
+     * The shape is determined by measuring the moments of the pixel
+     * centers about its centroid (cf. getCentroid)
+     */
+    geom::ellipses::Quadrupole getShape() const { return _spans->computeShape(); }
+
+    /**
+     * @brief Shift a Footprint by <tt>(dx, dy)</tt>
+     *
+     * @param dx - How much to move Footprint in column direction
+     * @param dy - How much to move in row direction
+     */
+    void shift(int dx, int dy);
+
+    /**
+    * @brief Shift a Footprint by a given extent
+    *
+    * @param d - ExtentI object which gives the dimensions the Footprint should be shifted
+    */
+    void shift(geom::ExtentI const & d) {shift(d.getX(), d.getY());}
+
+    /**
+     * @brief Return the Footprint's bounding box
+     */
+    geom::Box2I getBBox() const { return _spans->getBBox(); }
```

Comment:
Why return by value here and const reference below?  I think it's the same issues in play in both cases (I don't have a strong opinion, but we should be consistent).